### ISSUE TITLE
SSO on iOS 11 doesn't seem to work?

### ISSUE THREAD

**Issue by wei-lee** [openid/AppAuth-iOS#186]
Hi,

Maybe I have missed something but it seems I can't get SSO to work on iOS 11 with the library. Here is the environment I have for my app:

Xcode: Version 9.1 (9B55)
iOS version on the simulator: 11.1
identity provider: Keycloak with OpenConnectId

I also noticed that even for the same app, if I login the first time using SFAuthenticationSession, and then close the app and open it again immediately for the second time, when the SFAuthenticationSession is presented, I still have to enter my username/password again. This is different from the experience I have on Android and I would expect for the second time, I don't have to enter my username/password again and the SFAuthenticationSession should be closed straight away.

If I want to do SSO with multiple apps on iOS 11, is there anything else I need to set, other than make sure all apps are belong to the same group?

Thanks.

**Comment by WilliamDenniss** [comment]
Have you tried other identity providers, e.g. Google?  Also what about a real device? Sometimes the simulator can behave differently.

**Comment by wei-lee** [comment]
@WilliamDenniss Yes, I did test on a real device as well and it is the same result. What is also interesting is that if I don't stop the app, and re-login everything works as expected, I do not need to enter the username/password again. But once the app is killed, I have to enter my credential again. So it looks like one SFAuthenticationSession is only valid until the app is stopped?

**Comment by panva** [comment]
The use of persistent cookies is required for SFSVC and SFAS to work properly if i recall correctly.

Does your IdP maybe use transient cookies? We've seen reports in the past with our IdP that when users register they have to re-login the next time they launch the app, not after they login though. The difference between reg and login for us was the use of transient cookie for reg and login with a checkbox set a proper expiration on it.

**Comment by WilliamDenniss** [comment]
+1 to @panva's comment. Maybe we should document this.

**Comment by michaelhunziker** [comment]
@wei-lee: have you been able to work with this library in iOS 11?

**Comment by wei-lee** [comment]
@michaelhunziker Sorry for the delay, but no, I still couldn't get SSO work on iOS 11. Looks like I am not the only one that is having this problem: https://stackoverflow.com/questions/46569570/sfauthenticationsession-isnt-sharing-cookies-on-the-real-devices.

Looks like it is a problem with SFAuthenticationSession itself.

@WilliamDenniss @panva do you mind trying SSO on iOS and see if that works for you?

**Comment by wei-lee** [comment]
@panva I did choose "Remember Me" on the login screen and I am pretty sure that will use persistent cookie, but still no luck :(

**Comment by panva** [comment]
@wei-lee Yes, after setting the cookies to be session cookies and not transient SSO works.

**Comment by wei-lee** [comment]
I have done quite a lot of tests on iOS 11 and at this point, I have come to the conclusion that the SFAS sometimes doesn't work as expected.

Here is what I have done:

1. I cloned the example app from this repo
2. I created 2 apps from it, both of them using Google as the IDP and I installed both apps on the device.
3. I cleared all the web history and website data in Safari.
4. I logged into app A using my google account, and make sure I can read the user profile info.
5. I then tried to login to app B. When SFAS is opened, there is no my existing google account to pick and I was asked to login. For now, I cancel the login window.
6. I go to Safari and go to `myaccount.google.com`. I can see that I didn't logged into google at all.
7. Now, I open the app A again, and I try to authenticate again. This time when the SFAS is opened, I can see that my account is remembered and I can pick it and complete the login process.
8. Now, if I go back to Safari, and refresh `myaccount.google.com`, I can suddenly see I am logged in in Safari
9. If I go to app B again, this time I can see my current google account and login without enter the username and password again.

Step 7 doesn't always work, sometimes I have to enter my credentials again.

@panva @WilliamDenniss Have you guys come across anything similar before? I think in order to reproduce this, you have to make sure remove the existing cookies from Safari first. It looks like SFAS doesn't always sync the cookies from SFAS back to Safari.

I tried with Keycloak and I see the same behaviour. Sometimes the SFAS syncs the session back to Safari, and when this happens, the other apps can perform SSO. But sometimes this doesn't happen and SSO doesn't work in other apps.

This is not a bug with AppAuth itself, I just want to make sure I am not the only one seeing this issue. If this is the case, I think it should be documented the SSO doesn't always work.

**Comment by panva** [comment]
@wei-lee there were reports of the cookie sync having a bit of a delay, some reported at around ~10 seconds, this was on iOS 10 and SFSVC tho.

**Comment by wei-lee** [comment]
@panva I think it has become worse in iOS 11....

**Comment by WilliamDenniss** [comment]
So what we know is 1) that you MUST be using persistent cookies in order to get SSO on iOS 2) with persistent cookies, SSO is possible but there is a long running known-issue related to syncing delays between SFAS and the Safari cookie jar.

Do we know if iOS is eventually consistent with the cookies syncing? E.g. if you wait a certain amount of time, will it eventually sync the state? Or are there cases where it never sync's?

I strongly encourage everyone encountering this issue to [file a bug to Apple](https://developer.apple.com/bug-reporting/). We can keep this issue open here as a way to track this platform bug for AppAuth users, but be aware that no direct action will result from this AppAuth issue. The only way to get this resolved is to communicate the issue directly to Apple, such as via their bug reporting system.

When you file a bug, feel free to also file it at http://www.openradar.me/ so we can all view it. If you do, put the open radar link in this thread.

**Comment by CraigLn** [comment]
I've had an open issue logged with Apple with what I believe might be the same issue you're seeing: http://www.openradar.me/radar?id=5036182937272320

**Comment by guylando** [comment]
I found out the problem of why account is not shown in accounts chooser (at least after upgrading the ios sdk to SFAuthenticationSession, I don't know if it happened from same reason before upgrade):
The google sso authentication accounts are saved in safari data and if you clear safari cache then the accounts are removed. This can be easily reproduced by logging in, in one app (such as yelp, tripadvisor and others) and then you WILL see accounts in other apps on google login. Then clean safari cache and offline data and then you will not see the accounts in other apps on google login.

**Comment by jassicamark1991** [comment]
If you, instead of clicking on the icon in the Azure portal, simply navigate to your atlassian cloud URL, then instead of clicking on sign in with Microsoft, type your email, as if you are going to use password credentials - what happens then? If you are not redirected to Azure then - your SAML SSO hasn't been setup correctly.
